web/docs/glossary.md

```diff
@@ -0,0 +1,170 @@
+
+A connector is either a *loader* that acquires bytes from a resource, or a
+*saver* that sends bytes to a resource. Example connectors are
+[`file`](./connectors/file.md), [`kafka`](./connectors/kafka.md), and
+[`nic`](./connectors/nic.md).
+
+- See all available [connectors](./connectors.md)
+
+## Format
+
+Translates between bytes and events.
+
+A format is either a *parser* that converts bytes to events, or a *printer*
+that converts events to bytes. Example formats are [`json`](./formats/json.md),
+[`cef`](./formats/cef), and [`pcap`](./formats/pcap.md).
+
+- See all available [formats](./formats.md)
+
+## Index
+
+Optional data structures for accelerating historical queries.
+
+Tenzir has *sparse* and *dense* indexes. Sparse indexes live in memory and point
+to [partitions](#partition), whereas dense indexes live within a partition and
+point to individual rows within the partition [store](#store).
+
+- [Configure catalog and partition indexes](./setup-guides/tune-performance/README.md#configure-catalog-and-partition-indexes)
+
+## Node
+
+A host for [pipelines](#pipeline) and storage reachable over the network.
+
+The `tenzir-node` binary starts a node in a dedicated server process that
+listens on TCP port 5158.
+
+- [Deploy a node](./setup-guides/deploy-a-node/README.md)
+- [Use the app](./setup-guides/use-the-app/README.md) to manage a node
+- Use the [REST API](./rest-api.md) to manage a node.
+- [Import into a node](./user-guides/import-into-a-node.md)
+- [Export from a node](./user-guides/export-from-a-node.md)
+
+## Metrics
+
+Runtime statistics about pipeline and node execution.
+
+- [Collect metrics](./setup-guides/collect-metrics.md)
+- See all available [metrics](./metrics.md) details
+
+## Operator
+
+The atomic building block of a [pipeline](#pipeline).
+
+An operator is a [source](#source), [transformation](#transformation), or
+[sink](#sink)
+
+## Partition
+
+The horizontal scaling unit of the storage attached to a [node](#node).
+
+A partition contains a [store](#store) and optionally a set of indexes.
+
+## Pipeline
+
+Combines a set of [operators](#operator) into a dataflow graph.
+
+- Understand the [pipeline language](./language/pipelines.md)
+- [Run a pipeline](./user-guides/run-a-pipeline/README.md)
+
+## Platform
+
+Control plane for nodes and pipelines, accessible through [app](#app) at
+[app.tenzir.com](https://app.tenzir.com).
+
+- [Use the app](./setup-guides/use-the-app/README.md)
+
+## Schema
+
+A named record type describing the top-level structure of a data frame.
+
+[Schemas](./data-model/schemas.md) 
+
+## Sink
+
+An [operator](#operator) consuming input, without producing any output.
+
+- See all available [sinks](./operators/sinks/README.md)
+
+## Source
+
+An [operator](#operator) producing output, without consuming any input.
+
+- See all available [sources](./operators/sources/README.md)
+
+## Store
+
+The raw data in a [partition](#partition) in Parquet or Feather.
```

Comment:
The store is more than just the raw data itself. It is an abstraction that provides access to the data underlying a partition.

Mentioning Parquet and Feather here is also a bit of an implementation detail that is already covered by the first bullet.

Reply:
I'm going off this diagram: https://docs.tenzir.com/next/setup-guides/tune-performance#storage-engine. There, the store is _inside_ the partition and abstracts away the raw storage. Please make a concrete suggestion if you like it phrased differently.

> Mentioning Parquet and Feather here is also a bit of an implementation detail that is already covered by the first bullet.

I don't follow. Where is it covered? It's not a detail, actually, it's critical for gaining trust in the system by understanding what our abstractions map to. This is precisely the missing link that allows people to map our concepts to existing things out there.

Reply:
The store is not just the raw data, but also the mechanism to access it, and to scan it with an expression as a filter. It also contains metadata associated with the data, e.g., type information or the import timestamp.

I propose this:

```
A mechanism to read and write data belonging to a [partition](#partition). Every partition contains exactly one store.

Tenzir ships with [Apache Feather V2](link) and [Apache Parquet](link) stores.
```

Reply:
> The store is not just the raw data, but also the mechanism to access it, and to scan it with an expression as a filter.

I think it should be clear that Tenzir can access its stores efficiently. That's an implementation detail and doesn't help understanding the term.

> It also contains metadata associated with the data, e.g., type information or the import timestamp.

Isn't this part of the partition? Otherwise it wouldn't be possible to read the store with a third-party tool, unless we rely on the fact that a store has a dedicated place for metadata, which is a strong assumption.

Reply:
I've slightly tweaked the description.

Reply:
> > The store is not just the raw data, but also the mechanism to access it, and to scan it with an expression as a filter.
> I think it should be clear that Tenzir can access its stores efficiently. That's an implementation detail and doesn't help understanding the term.

I'm not arguing about this from a point of efficiency. It's about correctness. A store to us is not the data itself, but rather the mechanism to read and write the data. Technically, the store doesn't even know where the actual data resides, as the partition is responsible for that.

> > It also contains metadata associated with the data, e.g., type information or the import timestamp.
> 
> Isn't this part of the partition? Otherwise it wouldn't be possible to read the store with a third-party tool, unless we rely on the fact that a store has a dedicated place for metadata, which is a strong assumption.

The partition only contains aggregate information about the metadata that is required to read data, but the actual metadata is kept directly in the stores. This is by design—reading Arrow metadata is optional, and for both Feather and Parquet not reading it is as simple as opting out when creating the respective reader. We do want the stores to be self-contained, though.

Reply:
We agreed on this in a call.